# Exploring eBay Car Sales Data
## Introduction
In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The aim of this project is to clean the data and analyze the included used car listings. You'll also become familiar with some of the unique benefits jupyter notebook provides for pandas.

Let's start by importing the libraries we need and reading the dataset into pandas.

In [3]:
import pandas as pd 
import numpy as np

autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [4]:
autos

dateCrawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  yearOfRegistration  \
0       privat   Angebot    480     test         NaN                1993   
1       privat   Angebot  18300     test       coupe                2011   
2       privat   Angebot   9800     test         suv                2004   
3       privat   Angebot   1500     test  kleinwagen                2001   
4       privat   Angebot   3600     test  kleinwagen                2008   
...        ...       ...    ...      ...         ...                 ...   
371523  privat   Angebot   2200     test         NaN                2005   
371524  privat   Angebot   1199     test      cabrio                2000   
371525  privat   Angebot   9200     test         bus                1996   
371526  privat   Angebot   3400     test       kombi                2002   
371527  privat   Angebot  28990  control   limousine                2013   

          gearbox  powerPS        model  kilometer  monthOfRegistration  \
0         manuell        0         golf     150000                    0   
1         manuell      190          NaN     125000                    5   
2       automatik      163        grand     125000                    8   
3         manuell       75         golf     150000                    6   
4         manuell       69        fabia      90000                    7   
...           ...      ...          ...        ...                  ...   
371523        NaN        0          NaN      20000                    1   
371524  automatik      101       fortwo     125000                    3   
371525    manuell      102  transporter     150000                    3   
371526    manuell      100         golf     150000                    6   
371527    manuell      320      m_reihe      50000                    8   

       fuelType           brand notRepairedDamage          dateCreated  \
0        benzin      volkswagen               NaN  2016-03-24 00:00:00   
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
0                  0       70435  2016-04-07 03:16:57  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
...              ...         ...                  ..

In [7]:
print(autos.info())
print('\n')
print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

### Observations
The dataset has 371528 rows and 20 columns

Column names use camelcase rather than snakecase. 

There are missing values for `vehicleType`, `gearbox`, `model`, `fuelType`, `notRepairedDamage`. 
Most columns are of type string except for some integer values such as `price` and `powerPS`

from a brief observation, we can see that there are dates and time values stored as strings in `lastSeen` and `dateCreated`, for example "2016-04-05 12:47:46  "

The car name is also in an unusual format, see for example "Jeep_Grand_Cherokee_"Overland"" It appears there is no common naming convention, 

Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.



In [8]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [15]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_PS', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']


In [16]:
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price ab_test vehicle_type  registration_year   gear_box  power_PS  model  \
0    480    test          NaN               1993    manuell         0   golf   
1  18300    test        coupe               2011    manuell       190    NaN   
2   9800    test          suv               2004  automatik       163  grand   
3   1500    test   kleinwagen               2001    manuell        75   golf   
4   3600    test   kleinwagen               2008    manuell        69  fabia   

   odometer_km  registration_month fuel_type       brand unrepaired_damage  \
0       150000                   0    benzin  volkswagen               NaN   
1       125000                   5    diesel        audi                ja   
2       125000                   8    diesel        jeep               NaN   
3       150000                   6    benzin  volkswagen              nein   
4        90000                   7    diesel       skoda              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00               0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00               0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00               0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00               0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00               0        60437  2016-04-06 10:17:21

Column names have been edited from camel case to snake case, this allows for easier viewing and understanding. Some columns have been renamed to better reflect their meaning. 

### Data Cleaning 
Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:

Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
Examples of numeric data stored as text which can be cleaned and converted.

In [23]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

       ab_test vehicle_type  registration_year gear_box       power_PS  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model    odometer_km  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count             299468               371528        371528.0  371528.00000   
unique                 2                  114             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq              263182                14450             NaN           NaN   
mean                 NaN                  NaN             0.0   50820.66764   
std                  NaN                  NaN             0.0   25799.08247   
min                  NaN                  NaN             0.0    1067.00000   
25%                  NaN                  NaN             0.0   30459.00000   
50%                  NaN                  NaN             0.0   49610.00000   
75%                  NaN                  NaN             0.0   71546.00000   
max                  NaN                  NaN             0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                  

The `nr_of_pictures` column has a min, mean and max value of 0 suggesting all values are 0 in this column. 
The `seller` values are mostly 'privat'
also, the `offer_type` values are mostly 'Angebot'. 

We will remove these columns . 


In [24]:
autos = autos.drop(["nr_of_pictures", 'seller', 'offer_type'], axis = 1)

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the odometer_km and price columns. Here's the steps we'll take:

Analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.

## Exploring Price

In [18]:
autos['price'].unique().shape

(5597,)

In [20]:
autos['price'].describe()

count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64

In [22]:
autos['price'].value_counts().sort_index(ascending = False)

2147483647        1
99999999         15
99000000          1
74185296          1
32545461          1
              ...  
4                 1
3                 8
2                12
1              1189
0             10778
Name: price, Length: 5597, dtype: int64

There are 10,778 cars with a price of 0. 

The maximum prices seem very over priced

In [25]:
autos['price'].value_counts().sort_index(ascending = False).head(20)

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
2795000        1
1600000        2
1300000        1
1250000        2
1234566        1
Name: price, dtype: int64

There are a number of listings with prices below 
0. There are also a small number of listings with very high values, including 14 at around or over $1 million.

Given that eBay is an auction site, there could legitimately be items where the opening bid is 
1 items, but remove anything above $350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [26]:
autos = autos[autos["price"].between(1, 351000)]
autos['price'].describe()

count    360635.000000
mean       5898.671956
std        8866.359669
min           1.000000
25%        1250.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

By removing outliers in price, we have removed approx. 10, 000 rows in our data. The size is now 360635x20

## Exploring Odometer KM

In [27]:
autos['odometer_km'].unique().shape

(13,)

In [28]:
autos['odometer_km'].describe()

count    360635.000000
mean     125675.364288
std       39818.609118
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.

## Exploring Dates

Extracting the day from the YYY-MM-DD HH:MM:SS formats of all three date columns 

In [33]:
autos['date_crawled'] = autos['date_crawled'].str[:10]
autos['ad_created'] = autos['ad_created'].str[:10]
autos['last_seen'] = autos['last_seen'].str[:10]

print(autos['date_crawled'])

0         2016-03-24
1         2016-03-24
2         2016-03-14
3         2016-03-17
4         2016-03-31
             ...    
371523    2016-03-14
371524    2016-03-05
371525    2016-03-19
371526    2016-03-20
371527    2016-03-07
Name: date_crawled, Length: 360635, dtype: object


In [40]:
autos['date_crawled'].value_counts(normalize = True, dropna = False).sort_index(ascending = True)


2016-03-05    0.025547
2016-03-06    0.014483
2016-03-07    0.035657
2016-03-08    0.033469
2016-03-09    0.034115
2016-03-10    0.032645
2016-03-11    0.032773
2016-03-12    0.036242
2016-03-13    0.015783
2016-03-14    0.036330
2016-03-15    0.033424
2016-03-16    0.030205
2016-03-17    0.031647
2016-03-18    0.013119
2016-03-19    0.035271
2016-03-20    0.036400
2016-03-21    0.035682
2016-03-22    0.032493
2016-03-23    0.032002
2016-03-24    0.029914
2016-03-25    0.032800
2016-03-26    0.031974
2016-03-27    0.030227
2016-03-28    0.035063
2016-03-29    0.034126
2016-03-30    0.033535
2016-03-31    0.031872
2016-04-01    0.034145
2016-04-02    0.035094
2016-04-03    0.038812
2016-04-04    0.037628
2016-04-05    0.012780
2016-04-06    0.003128
2016-04-07    0.001617
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

In [41]:
autos['ad_created'].value_counts(normalize = True, dropna = False).sort_index(ascending = True)


2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
                ...   
2016-04-03    0.039001
2016-04-04    0.037736
2016-04-05    0.011613
2016-04-06    0.003119
2016-04-07    0.001553
Name: ad_created, Length: 114, dtype: float64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [44]:
autos['last_seen'].value_counts(normalize = True, dropna = False).sort_index(ascending = True)


2016-03-05    0.001264
2016-03-06    0.004098
2016-03-07    0.005202
2016-03-08    0.007939
2016-03-09    0.009824
2016-03-10    0.011460
2016-03-11    0.012955
2016-03-12    0.023240
2016-03-13    0.008410
2016-03-14    0.012176
2016-03-15    0.016324
2016-03-16    0.016418
2016-03-17    0.028699
2016-03-18    0.006888
2016-03-19    0.016330
2016-03-20    0.019884
2016-03-21    0.020026
2016-03-22    0.020508
2016-03-23    0.018015
2016-03-24    0.019163
2016-03-25    0.019000
2016-03-26    0.015958
2016-03-27    0.016721
2016-03-28    0.022189
2016-03-29    0.023284
2016-03-30    0.023725
2016-03-31    0.024243
2016-04-01    0.023897
2016-04-02    0.024967
2016-04-03    0.025308
2016-04-04    0.025536
2016-04-05    0.126962
2016-04-06    0.218950
2016-04-07    0.130437
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [45]:
autos['registration_year'].describe()

count    360635.000000
mean       2004.433133
std          81.016977
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

## Dealing with Incorrect Registration Year Data

Because a car can't be first registered before the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

One option is to remove the listings with these values. Let's determine what percentage of our data has invalid values in this column:

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [54]:
(-autos['registration_year'].between(1900, 2016)).sum() / autos.shape[0]

0.038751091824143526

4% of our data contains cars registered before 1900 or after their ad creation date, we will remove these rows.

In [55]:
autos = autos[autos['registration_year'].between(1900, 2016)]
autos['registration_year'].value_counts(normalize = True).head(10)

2000    0.066699
1999    0.063552
2005    0.062669
2006    0.057708
2001    0.056955
2003    0.056557
2004    0.056173
2002    0.054290
2007    0.050499
1998    0.049691
Name: registration_year, dtype: float64

It appears that most of the vehicles were first registered in the past 20 years.

## Exploring price by brand

In [61]:
print(autos['brand'].unique())



['volkswagen' 'audi' 'jeep' 'skoda' 'bmw' 'peugeot' 'ford' 'mazda'
 'nissan' 'renault' 'mercedes_benz' 'seat' 'honda' 'fiat' 'opel' 'mini'
 'smart' 'hyundai' 'sonstige_autos' 'alfa_romeo' 'subaru' 'volvo'
 'mitsubishi' 'kia' 'suzuki' 'lancia' 'porsche' 'citroen' 'toyota'
 'chevrolet' 'dacia' 'daihatsu' 'trabant' 'chrysler' 'jaguar' 'daewoo'
 'rover' 'saab' 'land_rover' 'lada']


In [65]:
autos['brand'].value_counts(normalize = True)

volkswagen        0.211700
bmw               0.109871
opel              0.106410
mercedes_benz     0.096841
audi              0.089543
ford              0.068918
renault           0.047516
peugeot           0.030153
fiat              0.025691
seat              0.018661
skoda             0.015687
mazda             0.015384
smart             0.014331
citroen           0.013950
nissan            0.013598
toyota            0.012932
hyundai           0.009972
sonstige_autos    0.009493
mini              0.009384
volvo             0.009147
mitsubishi        0.008236
honda             0.007532
kia               0.006915
suzuki            0.006364
alfa_romeo        0.006309
porsche           0.006211
chevrolet         0.005022
chrysler          0.003863
dacia             0.002495
jeep              0.002192
land_rover        0.002166
daihatsu          0.002161
subaru            0.002117
jaguar            0.001734
saab              0.001465
daewoo            0.001457
trabant           0.001408
l

German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [72]:
brand_counts = autos['brand'].value_counts(normalize = True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [81]:
brand_mean = {}

for br in common_brands:
    brands = autos[autos['brand'] == br]
    brand_mean[br] = int(brands['price'].mean())
    
brand_mean

{'volkswagen': 5400,
 'bmw': 8449,
 'opel': 2971,
 'mercedes_benz': 8551,
 'audi': 9086,
 'ford': 3696}

Of the top 5 brands, there is a distinct price gap:

Audi, BMW and Mercedes Benz are more expensive
Ford and Opel are less expensive
Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.

## Exploring Mileage

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:

it's difficult to compare more than two aggregate series objects if we want to extend to more columns
we can't compare more than a few rows from each series object
we can only sort by the index (brand name) of both series objects so we can easily make visual comparisons

In [83]:
bmp_series = pd.Series(brand_mean)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           5400
bmw                  8449
opel                 2971
mercedes_benz        8551
audi                 9086
ford                 3696

In [87]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos['brand'] == brand]
    mean_mileage = brand_only['odometer_km'].mean()
    brand_mean_mileage[brand] = int(mean_mileage)
    
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending = False)
mean_price = pd.Series(brand_mean).sort_values(ascending = False)


In [89]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132800
mercedes_benz        130572
audi                 129443
opel                 128722
volkswagen           128386
ford                 123662

In [90]:
brand_info["mean_price"] = mean_price
brand_info

mean_mileage  mean_price
bmw                  132800        8449
mercedes_benz        130572        8551
audi                 129443        9086
opel                 128722        2971
volkswagen           128386        5400
ford                 123662        3696

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.